source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -6811,10 +6843,124 @@ export class AdaptiveCard extends ContainerWithActions {
         if (fallbackElement) {
             this._fallbackCard = new AdaptiveCard();
             this._fallbackCard.addItem(fallbackElement);
-        }
+		}
+		
+		// uncomment line below to enable sending of telemetry
+		// ACLogger.getOrCreate().configureCustomProviders( { add providers here } );
+		ACLogger.getOrCreate().configureCustomProviders(new ConsoleProvider());
+
+		if (ACLogger.getOrCreate().isTelemetryEnabled()) {
+			this.renderCardTelemetry(json);
+		}
 
         super.parse(json, errors);
-    }
+	}
+	
+	/**
+	 * A method that creates an IACLogger and enables event logging to the respective providers.
+	 * 
+	 * @param json - The JSON object that is being parsed
+	 */
+	private renderCardTelemetry(json: any): void {
```

Comment:
Hi David,

I agree with your sentiment that another approach may be more ideal. I initially took this approach as to not add too many modifications to existing element classes as well as allow for easy tracking of only logging cards that have a rating input. However, reconsidering this, I see that while the former issue would still be valid, the impact could be minimized by adding a couple lines of code to each element’s internalRender() method to achieve the same effect. This would also ostensibly handle the Action edge cases as well as logging new elements. The latter issue is solved with the GUIDHelper keeping track of rating/no rating in parse().

I will explore this option of adding to internalRender() and update this commit with my findings. Please let me know what you think of this approach.

Reply:
Hi @dclaux ,

I prototyped a version in which logging calls are now made in each element's render() or internalRender() methods, adding just a couple lines of code at the beginning of each method to achieve this. This now solves your concerns of both edge cases and if/switch statements, as well as @shalinijoshi19 's concerns with extra bloat with the switch statements and conditional logic. Some things to consider if we go with this approach:

- Would it be more ideal if we created a telemetry wrapper class for each element class to house these new hooks? This would facilitate modularity but would add some extra bloat as well as add some extra dependencies (e.g. ColumnWithTelemetry would now extend ContainerWithTelemetry rather than Container). I am also unsure of how this would affect existing code - I can see that maybe I could add a new TypeRegistryWithTelemetry class and use this if telemetry metrics are desired, but I need to look more into this.
- One exception I found was in Facts due to the lack of explicit render() method. However, I see that this could be solved by logging within the parse() method of FactSet.

Reply:
This sounds like it's going in the right direction, but I would have to see your code to have a better idea.